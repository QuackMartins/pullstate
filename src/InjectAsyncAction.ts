import React from "react";
import {
  IAsyncActionBeckonOptions,
  IAsyncActionWatchOptions,
  IOCreateAsyncActionOutput,
  TPullstateAsyncBeckonResponse,
  TPullstateAsyncWatchResponse,
} from "./async-types";

export enum EAsyncActionInjectType {
  WATCH = "watch",
  BECKON = "beckon",
}

interface IPropsInjectAsyncActionBase<A, R, T extends string> {
  action: IOCreateAsyncActionOutput<A, R, T>;
  args?: A;
}

export interface IPropsInjectAsyncActionBeckon<A = any, R = any, T extends string = string>
  extends IPropsInjectAsyncActionBase<A, R, T> {
  type: EAsyncActionInjectType.BECKON;
  options?: IAsyncActionBeckonOptions;
  children: (response: TPullstateAsyncBeckonResponse<R, T>) => React.ReactElement;
}

export interface IPropsInjectAsyncActionWatch<A = any, R = any, T extends string = string>
  extends IPropsInjectAsyncActionBase<A, R, T> {
  type: EAsyncActionInjectType.WATCH;
  children: (response: TPullstateAsyncWatchResponse<R, T>) => React.ReactElement;
  options?: IAsyncActionWatchOptions;
}

export type TInjectAsyncActionProps = IPropsInjectAsyncActionBeckon | IPropsInjectAsyncActionWatch;

export function InjectAsyncAction(
  props: TInjectAsyncActionProps
): React.ReactElement {
  if (props.type === EAsyncActionInjectType.BECKON) {
    const response = props.action.useBeckon(props.args, props.options);
    return props.children(response);
  }

  const response = props.action.useWatch(props.args, props.options);
  return props.children(response);
}
