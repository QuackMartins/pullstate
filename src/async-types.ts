import { IPullstateAllStores } from "./PullstateCore";
import { TUpdateFunction } from "./Store";

type TPullstateAsyncUpdateListener = () => void;

// [ started, finished, result, updating ]
export type TPullstateAsyncWatchResponse<R = any, T extends string = string> = [
  boolean,
  boolean,
  TAsyncActionResult<R, T>,
  boolean
];

// [ started, finished, result, updating, postActionResult ]
export type TPullstateAsyncResponseCacheFull<R, T extends string> = [
  boolean,
  boolean,
  TAsyncActionResult<R, T>,
  boolean,
  TAsyncActionResult<R, T> | true | null
];
// [finished, result, updating]
export type TPullstateAsyncBeckonResponse<R = any, T extends string = string> = [
  boolean,
  TAsyncActionResult<R, T>,
  boolean
];
// [result]
export type TPullstateAsyncRunResponse<R = any, T extends string = string> = Promise<
  TAsyncActionResult<R, T>
>;

export interface IPullstateAsyncResultState {
  [key: string]: TPullstateAsyncWatchResponse<any, string>;
}

export interface IPullstateAsyncActionOrdState {
  [key: string]: number;
}

export enum EAsyncEndTags {
  THREW_ERROR = "THREW_ERROR",
  RETURNED_ERROR = "RETURNED_ERROR",
  UNFINISHED = "UNFINISHED",
}

interface IAsyncActionResultBase<T extends string> {
  message: string;
  tags: (EAsyncEndTags | T)[];
}

export interface IAsyncActionResultPositive<R, T extends string> extends IAsyncActionResultBase<T> {
  error: false;
  payload: R;
}

export interface IAsyncActionResultNegative<T extends string> extends IAsyncActionResultBase<T> {
  error: true;
  payload: null;
}

export type TAsyncActionResult<R, T extends string> =
  | IAsyncActionResultPositive<R, T>
  | IAsyncActionResultNegative<T>;

// Order of new hook functions:

// shortCircuitHook = ({ args, stores }) => cachable response | false     - happens only on uncached action
// cacheBreakHook = ({ args, stores, result }) => true | false            - happens only on cached action
// postActionHook = ({ args, result, stores }) => void | new result       - happens on all actions, after the async / short circuit has resolved
// ----> postActionHook potentially needs a mechanism which allows it to run only once per new key change (another layer caching of some sorts expiring on key change)

export type TPullstateAsyncShortCircuitHook<A, R, T extends string, S extends IPullstateAllStores> = (
  inputs: {
    args: A;
    stores: S;
  }
) => TAsyncActionResult<R, T> | false;

export type TPullstateAsyncCacheBreakHook<A, R, T extends string, S extends IPullstateAllStores> = (
  inputs: {
    args: A;
    result: TAsyncActionResult<R, T>,
    stores: S;
  }
) => boolean;

export enum EPostActionContext {
  WATCH_HIT_CACHE = "WATCH_HIT_CACHE",
  BECKON_HIT_CACHE = "BECKON_HIT_CACHE",
  RUN_HIT_CACHE = "RUN_HIT_CACHE",
  SHORT_CIRCUIT = "SHORT_CIRCUIT",
  DIRECT_RUN = "DIRECT_RUN",
  BECKON_RUN = "BECKON_RUN",
}

export type TPullstateAsyncPostActionHook<A, R, T extends string, S extends IPullstateAllStores> = (
  inputs: {
    args: A;
    result: TAsyncActionResult<R, T>,
    stores: S;
    context: EPostActionContext;
  }
) => void;

export interface IAsyncActionBeckonOptions {
  ssr?: boolean;
  postActionEnabled?: boolean;
  cacheBreakEnabled?: boolean;
}

export interface IAsyncActionWatchOptions extends IAsyncActionBeckonOptions {
  initiate?: boolean;
}

export interface IAsyncActionRunOptions<S extends IPullstateAllStores = any> {
  treatAsUpdate?: boolean;
  ignoreShortCircuit?: boolean;
  respectCache?: boolean;
  _asyncCache?: IPullstateAsyncCache;
  _stores?: S;
}

export interface IAsyncActionGetCachedOptions {
  checkCacheBreak?: boolean;
}

export interface IGetCachedResponse<R, T extends string> {
  started: boolean;
  finished: boolean;
  result: TAsyncActionResult<R, T>;
  updating: boolean;
  existed: boolean;
  cacheBreakable: boolean;
}

export type TAsyncActionBeckon<A, R, T extends string> = (
  args?: A,
  options?: IAsyncActionBeckonOptions
) => TPullstateAsyncBeckonResponse<R, T>;
export type TAsyncActionWatch<A, R, T extends string> = (
  args?: A,
  options?: IAsyncActionWatchOptions
) => TPullstateAsyncWatchResponse<R, T>;
export type TAsyncActionRun<A, R, T extends string> = (
  args?: A,
  options?: IAsyncActionRunOptions
) => TPullstateAsyncRunResponse<R, T>;
export type TAsyncActionClearCache<A> = (args?: A) => void;
export type TAsyncActionClearAllCache = () => void;
export type TAsyncActionSetCached<A, R, T extends string> = (args: A, result: TAsyncActionResult<R, T>) => void;
export type TAsyncActionUpdateCached<A, R, T extends string> = (args: A, updater: TUpdateFunction<R>) => void;
export type TAsyncActionGetCached<A, R, T extends string> = (args?: A, options?: IAsyncActionGetCachedOptions) => IGetCachedResponse<R, T>;
export type TAsyncActionDelayedRun<A> = (
  args?: A,
  options?: IAsyncActionRunOptions & { delay: number, clearOldRun?: boolean, immediateIfCached?: boolean }
) => () => void;

export interface IOCreateAsyncActionOutput<A = any, R = any, T extends string = string> {
  useBeckon: TAsyncActionBeckon<A, R, T>;
  useWatch: TAsyncActionWatch<A, R, T>;
  run: TAsyncActionRun<A, R, T>;
  delayedRun: TAsyncActionDelayedRun<A>;
  getCached: TAsyncActionGetCached<A, R, T>;
  setCached: TAsyncActionSetCached<A, R, T>;
  updateCached: TAsyncActionUpdateCached<A, R, T>;
  clearCache: TAsyncActionClearCache<A>;
  clearAllCache: TAsyncActionClearAllCache;
}

export interface IPullstateAsyncCache {
  results: IPullstateAsyncResultState;
  listeners: {
    [key: string]: {
      [watchId: string]: TPullstateAsyncUpdateListener;
    };
  };
  actions: {
    [key: string]: () => Promise<TAsyncActionResult<any, string>>;
  };
  actionOrd: IPullstateAsyncActionOrdState;
}

export type TPullstateAsyncAction<A, R, T extends string, S extends IPullstateAllStores> = (
  args: A,
  stores: S
) => Promise<TAsyncActionResult<R, T>>;

export interface ICreateAsyncActionOptions<A, R, T extends string, S extends IPullstateAllStores> {
  clientStores?: S;
  shortCircuitHook?: TPullstateAsyncShortCircuitHook<A, R, T, S>;
  cacheBreakHook?: TPullstateAsyncCacheBreakHook<A, R, T, S>;
  postActionHook?: TPullstateAsyncPostActionHook<A, R, T, S>;
}
