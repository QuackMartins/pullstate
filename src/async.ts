import { IPullstateAllStores, PullstateContext } from "./PullstateCore";
const shallowEqual = require("fbjs/lib/shallowEqual");
import { useContext, useEffect, useState } from "react";

type TPullstateAsyncUpdateListener = () => void;

// result state = [ started, finished, resp, updating ]
export type TPullstateAsyncWatchResponse<R> = [boolean, boolean, null | R, boolean];
export type TPullstateAsyncBeckonResponse<R> = [boolean, null | R, boolean];
export type TPullstateAsyncRunResponse<R> = Promise<[null | R]>;
// export interface IPullstateAsyncResponse<R> {
//   started: boolean;
//   finished: boolean;
//   updating: boolean;
//   result: R;
// }

export interface IPullstateAsyncState {
  [key: string]: TPullstateAsyncWatchResponse<any>;
}

export interface IPullstateAsyncCache {
  results: IPullstateAsyncState;
  listeners: {
    [key: string]: TPullstateAsyncUpdateListener[];
  };
  actions: {
    [key: string]: () => Promise<any>;
  };
}

export type TPullstateAsyncAction<A, R, S extends IPullstateAllStores> = (args: A, stores: S) => Promise<R>;

export interface IAsyncActionWatchOptions {
  initiate?: boolean;
}

export interface IAsyncActionRunOptions {
  treatAsUpdate?: boolean;
}

type TAsyncActionBeckon<A, R> = (args?: A) => TPullstateAsyncBeckonResponse<R>;
type TAsyncActionWatch<A, R> = (args?: A, options?: IAsyncActionWatchOptions) => TPullstateAsyncWatchResponse<R>;
type TAsyncActionRun<A, R> = (args?: A, options?: IAsyncActionRunOptions) => TPullstateAsyncRunResponse<R>;
type TAsyncActionClearCache<A> = (args?: A) => void;

export interface IOCreateAsyncActionOutput<A, R> {
  beckon: TAsyncActionBeckon<A, R>;
  watch: TAsyncActionWatch<A, R>;
  run: TAsyncActionRun<A, R>;
  clearCache: TAsyncActionClearCache<A>;
}

export const clientAsyncCache: IPullstateAsyncCache = {
  listeners: {},
  results: {},
  actions: {},
};

let asyncCreationOrdinal = 0;

export function keyFromObject(json: any): string {
  if (json == null) {
    return `${json}`;
  }

  let prefix = "";

  for (const key of Object.keys(json).sort()) {
    prefix += key;

    if (typeof json[key] == null) {
      prefix += JSON.stringify(json[key]);
    } else if (typeof json[key] === "string") {
      prefix += `~${json[key]}~`;
    } else if (typeof json[key] === "boolean" || typeof json[key] === "number") {
      prefix += json[key];
    } else {
      prefix += keyFromObject(json[key]);
    }
  }

  return prefix;
}

function createKey(ordinal, args: any) {
  return `${ordinal}-${keyFromObject(args)}`;
}

function notifyListeners(key: string) {
  console.log(`Notifying (${clientAsyncCache.listeners[key].length}) listeners for key: ${key}`);
  for (const listener of clientAsyncCache.listeners[key]) {
    listener();
  }
}

export function createAsyncAction<A, R, S extends IPullstateAllStores = IPullstateAllStores>(
  action: TPullstateAsyncAction<A, R, S>,
  defaultArgs: A = {} as A,
  clientStores: S = {} as S
): IOCreateAsyncActionOutput<A, R> {
  const ordinal: number = asyncCreationOrdinal++;
  const onServer: boolean = typeof window === "undefined";
  // console.log(`Creating async action with ordinal: ${ordinal} - action name: ${action.name}`);

  const watch: TAsyncActionWatch<A, R> = (args = defaultArgs, { initiate = false }: IAsyncActionWatchOptions = {}) => {
    const key = createKey(ordinal, args);
    let shouldUpdate = true;

    const cache: IPullstateAsyncCache = onServer ? useContext(PullstateContext)._asyncCache : clientAsyncCache;
    const stores = onServer ? (useContext(PullstateContext).stores as S) : clientStores;

    function checkKeyAndReturnResponse(key: string): TPullstateAsyncWatchResponse<R> {
      if (cache.results.hasOwnProperty(key)) {
        // console.log(`Pullstate Async: [${key}] Already been run - do nothing`);
        return cache.results[key];
        // return [true, true, null, false];
      }

      // console.log(`Pullstate Async: [${key}] has no results yet`);

      // check if it is already pending as an action
      if (!cache.actions.hasOwnProperty(key)) {
        if (initiate) {
          // queue (on server) or start the action now (on client)
          cache.actions[key] = () => action(args, stores);

          if (!onServer) {
            cache.actions[key]()
              .then(resp => {
                cache.results[key] = [true, true, resp, false];
              })
              .catch(() => {
                cache.results[key] = [true, true, null, false];
              })
              .then(() => {
                delete cache.actions[key];
                notifyListeners(key);
              });
          }
        } else {
          return [false, false, null, false];
        }
      }

      return [true, false, null, false];
    }

    // only listen for updates when on client
    if (!onServer) {
      function onAsyncStateChanged() {
        if (shouldUpdate && !shallowEqual(response, cache.results[key])) {
          setResponse(checkKeyAndReturnResponse(key));
        }
      }

      useEffect(() => {
        if (!cache.listeners.hasOwnProperty(key)) {
          cache.listeners[key] = [];
        }

        // console.log(`Adding listener for key: ${key}`);
        cache.listeners[key].push(onAsyncStateChanged);

        return () => {
          shouldUpdate = false;
          // console.log(`Removing listener for key: ${key}`);
          cache.listeners[key] = cache.listeners[key].filter(f => f !== onAsyncStateChanged);
        };
      }, [key]);
    }

    const [response, setResponse] = useState<TPullstateAsyncWatchResponse<R>>(() => {
      return checkKeyAndReturnResponse(key);
    });

    const [prevKey, setPrevKey] = useState<string>(key);

    if (prevKey !== key) {
      setPrevKey(key);
      setResponse(checkKeyAndReturnResponse(key));
    }

    return response;
  };

  // Same as watch - just initiated, so no need for "started" return value
  const beckon: TAsyncActionBeckon<A, R> = (args = defaultArgs) => {
    const result = watch(args, { initiate: true });
    return [result[1], result[2], result[3]];
  }

  const run: TAsyncActionRun<A, R> = async (args = defaultArgs, { treatAsUpdate = false }: IAsyncActionRunOptions = {}) => {
    const key = createKey(ordinal, args);

    const [prevFinished, prevResp] = clientAsyncCache.results[key] || [false, null];

    if (prevFinished && treatAsUpdate) {
      clientAsyncCache.results[key] = [true, true, prevResp, true];
    } else {
      clientAsyncCache.results[key] = [true, false, null, false];
    }

    notifyListeners(key);

    // for (const listener of clientAsyncCache.listeners[key]) {
    //   listener();
    // }

    try {
      const resp = await action(args, clientStores);
      clientAsyncCache.results[key] = [true, true, resp, false];
      notifyListeners(key);
      return [resp];
    } catch (e) {
      clientAsyncCache.results[key] = [true, true, null, false];
      notifyListeners(key);
      return [null];
    }
  };

  const clearCache: TAsyncActionClearCache<A> = (args = {} as A) => {
    const key = createKey(ordinal, args);
    delete clientAsyncCache.results[key];
    notifyListeners(key);
  };

  return {
    beckon,
    watch,
    run,
    clearCache,
  };
}
