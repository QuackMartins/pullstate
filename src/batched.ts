/*
import { clientStores, IPullstateAllStores } from "./PullstateCore";
// import { Store, TUpdateFunction } from "./Store";

/!*type TMultiStoreUpdateMap<S extends IPullstateAllStores> = { [K in keyof S]: (updater: TUpdateFunction<S[K] extends Store<infer T> ? T : any>) => void }

export function createAction<A extends any, S extends IPullstateAllStores = IPullstateAllStores>(
  batchUpdater: (
    update: TMultiStoreUpdateMap<S>,
    args: A
  ) => void,
  stores: S = clientStores.stores as S
): (args: A) => void {
  const update: TMultiStoreUpdateMap<S> = {} as TMultiStoreUpdateMap<S>;
  const updatedStores = new Set<string>();

  for (const store of Object.keys(stores)) {
    update[store as keyof S] = (updater) => {
      updatedStores.add(store);
      stores[store].batch(updater);
    };
  }

  return (args: A, updateMap = update, updatedSet = updatedStores) => {
    updatedSet.clear();
    batchUpdater(updateMap, args);
    for (const store of updatedSet) {
      stores[store].flushBatch();
    }
  }
}*!/

export function createBatchAction<A extends any>(batchedUpdates: (args: A) => void): (args: A) => void {
  return (args: A, stores: IPullstateAllStores = clientStores.stores) => {
    batchedUpdates(args);
    for (const store in Object.keys(stores)) {
      stores[store].flushBatch();
    }
  }
}
*/
