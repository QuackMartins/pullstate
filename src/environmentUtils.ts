export function isWebOrReactNative(): boolean {
  return typeof document !== "undefined" || (navigator?.product === "ReactNative");
}
