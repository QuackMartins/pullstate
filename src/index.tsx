import shallowEqual from "fbjs/lib/shallowEqual";
import immer from "immer";
import React, { useEffect, useState, useContext } from "react";

export type TPullstateUpdateListener = () => void;

// S = State
class Store<S = any> {
  private updateListeners: TPullstateUpdateListener[] = [];
  private currentState: S;
  private readonly initialState: S;

  constructor(initialState: S) {
    console.log(`Constructing ${this.constructor.name}`);
    this.currentState = initialState;
    this.initialState = initialState;
  }

  _getState(): S {
    return this.currentState;
  }

  _updateState(nextState: S) {
    this.currentState = nextState;
    this.updateListeners.forEach(listener => listener());
  }

  _addUpdateListener(listener: TPullstateUpdateListener) {
    this.updateListeners.push(listener);
  }

  _removeUpdateListener(listener: TPullstateUpdateListener) {
    this.updateListeners = this.updateListeners.filter(f => f !== listener);
  }

  update(updater: (state: S) => void) {
    update(this, updater);
  }
}

function update<S = any>(store: Store<S>, updater: (state: S) => void) {
  const currentState: S = store._getState();
  const nextState: S = immer(currentState as any, updater);
  if (nextState !== currentState) {
    store._updateState(nextState);
  }
}

// S = State
// SS = Sub-state
function useStoreState<S = any>(store: Store<S>): S;
function useStoreState<S = any, SS = any>(store: Store<S>, getSubState: (state: S) => SS): SS;
function useStoreState(store: Store, getSubState?: (state) => any): any {
  const [subState, setSubState] = useState<any | null>(null);
  let shouldUpdate = true;

  function onStoreUpdate() {
    const nextSubState = getSubState ? getSubState(store._getState()) : store._getState();
    if (shouldUpdate && !shallowEqual(subState, nextSubState)) {
      setSubState(nextSubState);
    }
  }

  useEffect(() => {
    store._addUpdateListener(onStoreUpdate);

    return () => {
      shouldUpdate = false;
      store._removeUpdateListener(onStoreUpdate);
    };
  });

  if (subState === null) {
    return getSubState ? getSubState(store._getState()) : store._getState();
  }

  return subState;
}

export interface IPropsInjectStoreState<S extends any = any, SS extends any = any> {
  store: Store<S>;
  getSubState?: (state: S) => SS;
  children: (output: SS) => React.ReactElement;
}

function InjectStoreState<S = any, SS = any>({
  store,
  children,
  getSubState = s => s as any,
}: IPropsInjectStoreState<S, SS>) {
  const state: SS = useStoreState(store, getSubState);
  return children(state);
}

export interface IPullstateAllStores {
  [storeName: string]: Store<any>;
}

const PullstateContext = React.createContext<IPullstateAllStores>({});

const PullstateProvider = <T extends IPullstateAllStores = IPullstateAllStores>({
  stores,
  children,
}: {
  stores: T;
  children?: any;
}) => {
  return <PullstateContext.Provider value={stores}>{children}</PullstateContext.Provider>;
};

let singleton: PullstateSingleton | null = null;

export class PullstateSingleton<T extends IPullstateAllStores = IPullstateAllStores> {
  private readonly allInitialStores: T = {} as T;

  constructor(allStores: T) {
    if (singleton !== null) {
      console.error(
        `Pullstate: createPullstate() - Should not be creating the core Pullstate class more than once! In order to re-use pull state, you need to call instantiate() on your already created object.`
      );
    }

    singleton = this;
    this.allInitialStores = allStores;
  }

  instantiate({
    hydrateState = null,
    reuseStores = false,
  }: { hydrateState?: any; reuseStores?: boolean } = {}): PullstateInstance<T> {
    if (reuseStores) {
      const instantiated = new PullstateInstance(this.allInitialStores);

      if (hydrateState != null) {
        instantiated.hydrateFromAllState(hydrateState);
      }

      return instantiated;
    }

    const newStores = {} as T;

    for (const storeName of Object.keys(this.allInitialStores)) {
      if (hydrateState == null) {
        newStores[storeName] = new Store(this.allInitialStores[storeName]._getState());
      } else if (hydrateState.hasOwnProperty(storeName)) {
        newStores[storeName] = new Store(hydrateState[storeName]);
      } else {
        newStores[storeName] = new Store(this.allInitialStores[storeName]._getState());
        console.warn(
          `Pullstate (instantiate): store [${storeName}] didn't hydrate any state (data was non-existent on hydration object)`
        );
      }
    }

    return new PullstateInstance(newStores);
  }

  useStores() {
    return useContext(PullstateContext) as T;
  }
}

class PullstateInstance<T extends IPullstateAllStores = IPullstateAllStores> {
  private readonly _stores: T = {} as T;

  constructor(allStores: T) {
    this._stores = allStores;
  }

  getAllState(): { [storeName: string]: any } {
    const allState = {};

    for (const storeName of Object.keys(this._stores)) {
      allState[storeName] = this._stores[storeName]._getState();
    }

    return allState;
  }

  get stores(): T {
    return this._stores;
  }

  hydrateFromAllState(allState: any) {
    for (const storeName of Object.keys(this._stores)) {
      if (allState.hasOwnProperty(storeName)) {
        this._stores[storeName]._updateState(allState[storeName]);
      } else {
        console.warn(`${storeName} didn't hydrate any state (data was non-existent on hydration object)`);
      }
    }
  }
}

function createPullstate<T extends IPullstateAllStores = IPullstateAllStores>(allStores: T) {
  return new PullstateSingleton(allStores);
}

function useStores<T extends IPullstateAllStores = {}>() {
  return useContext(PullstateContext) as T;
}

export { useStoreState, update, Store, InjectStoreState, PullstateProvider, useStores, createPullstate };
