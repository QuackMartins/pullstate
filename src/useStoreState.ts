const isEqual = require("fast-deep-equal");

// S = State
// SS = Sub-state
import { useCallback, useEffect, useRef, useState } from "react";
import { Store } from "./Store";

interface IUpdateRef {
  shouldUpdate: boolean;
  onStoreUpdate: () => void;
  getSubState: any;
  currentSubState: any;
}

function useStoreState<S = any>(store: Store<S>): S;
function useStoreState<S = any, SS = any>(store: Store<S>, getSubState: (state: S) => SS): SS;
function useStoreState(store: Store, getSubState?: (state) => any): any {
  const [subState, setSubState] = useState<any>(() =>
    getSubState ? getSubState(store.getRawState()) : store.getRawState()
  );

  const updateRef = useRef<IUpdateRef>({
    shouldUpdate: true,
    onStoreUpdate: null,
    getSubState,
    currentSubState: null,
  });

  updateRef.current.currentSubState = subState;
  updateRef.current.getSubState = getSubState;

  const onStoreUpdate = useCallback(() => {
    const nextSubState = updateRef.current.getSubState ? updateRef.current.getSubState(store.getRawState()) : store.getRawState();
    if (updateRef.current.shouldUpdate && !isEqual(updateRef.current.currentSubState, nextSubState)) {
      setSubState(nextSubState);
    }
  }, []);

  if (updateRef.current.onStoreUpdate === null) {
    updateRef.current.onStoreUpdate = onStoreUpdate;
    store._addUpdateListener(updateRef.current.onStoreUpdate);
  }

  useEffect(() => () => {
      updateRef.current.shouldUpdate = false;
      store._removeUpdateListener(onStoreUpdate);
  }, []);

  return subState;
}

export { useStoreState };
