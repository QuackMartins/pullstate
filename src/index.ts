import { useStoreState } from "./useStoreState";
import { Store, update } from "./Store";
import { InjectStoreState } from "./InjectStoreState";
import { createPullstateCore, IPullstateInstanceConsumable, PullstateProvider, useStores } from "./PullstateCore";
import { createAsyncAction, errorResult, successResult } from "./async";
import { EAsyncActionInjectType, InjectAsyncAction, TInjectAsyncActionProps } from "./InjectAsyncAction";
import { EAsyncEndTags, TPullstateAsyncAction } from "./async-types";

export {
  useStoreState,
  update,
  Store,
  InjectStoreState,
  PullstateProvider,
  useStores,
  createPullstateCore,
  createAsyncAction,
  successResult,
  errorResult,
  EAsyncEndTags,
  IPullstateInstanceConsumable,
  InjectAsyncAction,
  EAsyncActionInjectType,
  TInjectAsyncActionProps,
  TPullstateAsyncAction
};
