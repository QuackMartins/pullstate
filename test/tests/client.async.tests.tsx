import { Store } from "../../src/index";
import { useStoreState } from "../../src/useStoreState";
import { waitSeconds } from "./TestUtils";
import React, { useState } from "react";
import { createAsyncAction } from "../../src/async";
import { render } from "react-testing-library";
import ReactDOMServer from "react-dom/server";
const beautifyHtml = require("js-beautify").html;

const names = ["Paul", "Dave", "Michel"];
const userNames = ["lostpebble", "davej", "mweststrate"];
let currentUser = 0;

interface IUser {
  name: string;
  userName: string;
}

async function getNewUser(userId = -1): Promise<IUser> {
  currentUser = userId >= 0 ? userId : (currentUser + 1) % 3;

  await waitSeconds(1);

  return {
    name: names[currentUser],
    userName: userNames[currentUser],
  };
}

interface IUserStore {
  user: null | IUser;
}

const UserStore = new Store<IUserStore>({
  user: null,
});

const HydrateNewUserAction = createAsyncAction(async () => {
  const newUser = await getNewUser();
  UserStore.update(s => {
    s.user = newUser;
  });
  return true;
});

const GetUserAction = createAsyncAction(async ({ userId }) => {
  return await getNewUser(userId);
}, { userId: 0 });

const UninitatedUserAction = () => {
  const [userId, setUserId] = useState(0);
  const [started, finished, user] = GetUserAction.watch({ userId });

  return (
    <div>
      <span>{started ? finished ? `Got new user` : `Getting new user` : `Haven't initiated getting new user`}</span>
      {user !== null && (
        <div id="user-box">
          <h2>Hello, {user.name}</h2>
          <h3>aka: {user.userName}</h3>
        </div>
      )}
      {!started && <button id="uninitiated-get-user-button" onClick={() => GetUserAction.run({ userId })}>Initiate Get User for ID: {userId}</button>}
    </div>
  )
}

const InitiatedNextUser = () => {
  const user = useStoreState(UserStore, s => s.user);
  const [finished] = HydrateNewUserAction.beckon();

  return (
    <div>
      <span>{finished ? `User loaded` : `Loading user`}</span>
      {user !== null && (
        <div id="user-box">
          <h2>Hello, {user.name}</h2>
          <h3>aka: {user.userName}</h3>
        </div>
      )}
      <button disabled={(!finished)} id="initiated-get-user-button" onClick={() => HydrateNewUserAction.run({})}>
        Get next user
      </button>
    </div>
  )
}

const App = () => {
  return (
    <div>
      <h1>Async Test</h1>
      <InitiatedNextUser/>
      <UninitatedUserAction/>
    </div>
  );
};

describe("Async rendering", () => {
  it("renders our initial state", () => {
    const reactHtml = ReactDOMServer.renderToString(<App/>);
    expect(beautifyHtml(reactHtml)).toMatchSnapshot();
  });
});
