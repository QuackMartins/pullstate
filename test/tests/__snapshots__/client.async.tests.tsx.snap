// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async rendering renders our initial state 1`] = `
"<div data-reactroot=\\"\\">
    <h1>Async Test</h1>
    <div><span>Loading user</span><button disabled=\\"\\" id=\\"initiated-get-user-button\\">Get next user</button></div>
    <div><span>Haven&#x27;t initiated getting new user</span><button id=\\"uninitiated-get-user-button\\">Initiate Get User for ID:
            <!-- -->0</button></div>
</div>"
`;
